File: src/pdf/libqpdfselection.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const QRectF = @import("libqt6").QRectF;
const qclipboard_enums = @import("../libqclipboard.zig").enums;
const std = @import("std");

/// ### [Upstream resources](https://doc.qt.io/qt-6/qpdfselection.html)
pub const QPdfSelection = extern struct {
    /// ### [Upstream resources](https://doc.qt.io/qt-6/qpdfselection.html)
    ///
    /// The pointer to the underlying Qt C++ object
    ///
    ptr: QtC.QPdfSelection,

    pub const _is_QPdfSelection = {};

    /// ### DEPRECATED: Use `new` instead
    ///
    pub const New = new;

    /// Allocate a new QPdfSelection object in C++ memory
    ///
    /// ## Parameter(s):
    ///
    /// ` other: QPdfSelection `
    ///
    pub fn new(other: anytype) QPdfSelection {
        comptime _ = @TypeOf(other)._is_QPdfSelection;
        return .{ .ptr = qtc.QPdfSelection_new(@ptrCast(other.ptr)) };
    }

    /// ### DEPRECATED: Use `operatorAssign` instead
    ///
    pub const OperatorAssign = operatorAssign;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qpdfselection.html#operator-eq)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QPdfSelection `
    ///
    /// ` other: QPdfSelection `
    ///
    pub fn operatorAssign(self: QPdfSelection, other: anytype) void {
        comptime _ = @TypeOf(other)._is_QPdfSelection;
        qtc.QPdfSelection_OperatorAssign(@ptrCast(self.ptr), @ptrCast(other.ptr));
    }

    /// ### DEPRECATED: Use `swap` instead
    ///
    pub const Swap = swap;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qpdfselection.html#swap)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QPdfSelection `
    ///
    /// ` other: QPdfSelection `
    ///
    pub fn swap(self: QPdfSelection, other: anytype) void {
        comptime _ = @TypeOf(other)._is_QPdfSelection;
        qtc.QPdfSelection_Swap(@ptrCast(self.ptr), @ptrCast(other.ptr));
    }

    /// ### DEPRECATED: Use `isValid` instead
    ///
    pub const IsValid = isValid;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qpdfselection.html#isValid)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QPdfSelection `
    ///
    pub fn isValid(self: QPdfSelection) bool {
        return qtc.QPdfSelection_IsValid(@ptrCast(self.ptr));
    }

    /// ### DEPRECATED: Use `text` instead
    ///
    pub const Text = text;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qpdfselection.html#text)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QPdfSelection `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn text(self: QPdfSelection, allocator: std.mem.Allocator) []const u8 {
        var _str = qtc.QPdfSelection_Text(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("QPdfSelection.text: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `boundingRectangle` instead
    ///
    pub const BoundingRectangle = boundingRectangle;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qpdfselection.html#boundingRectangle)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QPdfSelection `
    ///
    pub fn boundingRectangle(self: QPdfSelection) QRectF {
        return .{ .ptr = qtc.QPdfSelection_BoundingRectangle(@ptrCast(self.ptr)) };
    }

    /// ### DEPRECATED: Use `startIndex` instead
    ///
    pub const StartIndex = startIndex;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qpdfselection.html#startIndex)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QPdfSelection `
    ///
    pub fn startIndex(self: QPdfSelection) i32 {
        return qtc.QPdfSelection_StartIndex(@ptrCast(self.ptr));
    }

    /// ### DEPRECATED: Use `endIndex` instead
    ///
    pub const EndIndex = endIndex;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qpdfselection.html#endIndex)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QPdfSelection `
    ///
    pub fn endIndex(self: QPdfSelection) i32 {
        return qtc.QPdfSelection_EndIndex(@ptrCast(self.ptr));
    }

    /// ### DEPRECATED: Use `copyToClipboard` instead
    ///
    pub const CopyToClipboard = copyToClipboard;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qpdfselection.html#copyToClipboard)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QPdfSelection `
    ///
    pub fn copyToClipboard(self: QPdfSelection) void {
        qtc.QPdfSelection_CopyToClipboard(@ptrCast(self.ptr));
    }

    /// ### DEPRECATED: Use `copyToClipboard1` instead
    ///
    pub const CopyToClipboard1 = copyToClipboard1;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qpdfselection.html#copyToClipboard)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QPdfSelection `
    ///
    /// ` mode: qclipboard_enums.Mode `
    ///
    pub fn copyToClipboard1(self: QPdfSelection, mode: i32) void {
        qtc.QPdfSelection_CopyToClipboard1(@ptrCast(self.ptr), @bitCast(mode));
    }

    /// ### DEPRECATED: Use `delete` instead
    ///
    pub const Delete = delete;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qpdfselection.html#dtor.QPdfSelection)
    ///
    /// Delete this object from C++ memory
    ///
    /// ## Parameter:
    ///
    /// ` self: QPdfSelection `
    ///
    pub fn delete(self: QPdfSelection) void {
        qtc.QPdfSelection_Delete(@ptrCast(self.ptr));
    }
};

File: src/pdf/libqpdfselection.cpp
#include <QList>
#include <QPdfSelection>
#include <QPolygonF>
#include <QRectF>
#include <QString>
#include <qpdfselection.h>
#include "libqpdfselection.h"
#include "libqpdfselection.hxx"

QPdfSelection* QPdfSelection_new(const QPdfSelection* other) {
    return new QPdfSelection(*other);
}

void QPdfSelection_OperatorAssign(QPdfSelection* self, const QPdfSelection* other) {
    self->operator=(*other);
}

void QPdfSelection_Swap(QPdfSelection* self, QPdfSelection* other) {
    self->swap(*other);
}

bool QPdfSelection_IsValid(const QPdfSelection* self) {
    return self->isValid();
}

libqt_list /* of QPolygonF* */ QPdfSelection_Bounds(const QPdfSelection* self) {
    QList<QPolygonF> _ret = self->bounds();
    // Convert QList<> from C++ memory to manually-managed C memory
    QPolygonF** _arr = static_cast<QPolygonF**>(malloc(sizeof(QPolygonF*) * (_ret.size())));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = new QPolygonF(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_string QPdfSelection_Text(const QPdfSelection* self) {
    auto _ret = self->text();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

QRectF* QPdfSelection_BoundingRectangle(const QPdfSelection* self) {
    return new QRectF(self->boundingRectangle());
}

int QPdfSelection_StartIndex(const QPdfSelection* self) {
    return self->startIndex();
}

int QPdfSelection_EndIndex(const QPdfSelection* self) {
    return self->endIndex();
}

void QPdfSelection_CopyToClipboard(const QPdfSelection* self) {
    self->copyToClipboard();
}

void QPdfSelection_CopyToClipboard1(const QPdfSelection* self, int mode) {
    self->copyToClipboard(static_cast<QClipboard::Mode>(mode));
}

void QPdfSelection_Delete(QPdfSelection* self) {
    delete self;
}
